text/0000-macro-at-most-once-rep.md

```diff
@@ -0,0 +1,166 @@
+
+Motivation
+----------
+
+There are two specific use cases in mind:
+
+1. Any macro rule with optional parts. Currently, you just have to write two rules and possibly have one "desugar" to the other.
+
+   ```rust
+   macro_rules! foo {
+     (do $b:block) => {
+       $b
+     }
+     (do $b1:block and $b2:block) => {
+       foo!($b1)
+       $b2
+     }
+   }
+   ```
+
+   Under this RFC, one would simply write:
+
+   ```rust
+   macro_rules! foo {
+     (do $b1:block $(and $b2:block)?) => {
+       $b1
+       $($b2)?
+     }
+   }
+   ```
+
+2. Trailing commas. It's kind of infuriating that the best way to make a rule tolerate trailing commas TMK is to create another identical rule that has a comma at the end:
+
+   ```rust
+   macro_rules! foo {
+     ($(pat),*,) => { foo!( $(pat),* ) };
+     ($(pat),*) => {
+       // do stuff
+     }
+   }
+   ```
+
+   or to allow multiple trailing commas:
+
+   ```rust
+   macro_rules! foo {
+     ($(pat),* $(,)*) => {
+       // do stuff
+     }
+   }
+   ```
+
+   Under this RFC, one would simply write:
+
+   ```rust
+   macro_rules! foo {
+     ($(pat),* $(,)?) => {
+       // do stuff
+     }
+   }
+   ```
+
+Guide-level explanation
+-----------------------
+
+In Rust macros, you specify some "rules" which define how the macro is used and what it transforms to. For each rule, there is a pattern and a body:
+
+```rust
+macro_rules! foo {
+  (pattern) => { body }
+}
+```
+
+The pattern portion is composed of zero or more subpatterns concatenated together. One possible subpattern is to repeat another subpattern some number of times. This is extremely useful when writing variadic macros (e.g. `println`):
+
+```rust
+macro_rules! println {
+  // Takes a variable number of arguments after the template
+  ($tempate:expr, $($args:expr),*) => { ... }
+}
+```
+which can be invoked like so:
+```rust
+println!("")           // 0 args
+println!("", foo)      // 1 args
+println!("", foo, bar) // 2 args
+...
+```
+
+The `*` in the pattern of this example indicates "0 or more repetitions". One can also use `+` for "at _least_ one repetition" or `?` for "at _most_ one repetition".
+
+In the body of a rule, one can specify to repeat some code for every occurence of the pattern in the invokation:
+
+```rust
+macro_rules! foo {
+  ($($pat:expr),*) => {
+    $(
+      println!("{}", $pat)
+    )* // Repeat for each `expr` passed to the macro
+  }
+}
+```
+
+The same can be done for `+` and `?`.
+
+The `?` operator is particularly useful for making macro rules with optional components in the invocation or for making macros tolerate trailing commas.
+
+Reference-level explanation
+---------------------------
+
+`?` is identical to `+` and `*` in use except that it represents "at most once" repetition. The implementation ought to be very similar to them. IIUC only the parser needs to change. I don't think it would be technically difficult to implement, nor do I think it would add much complexity to the compiler.
```

Comment:
Stylistic suggestions:
+ depersonalize IIUC (and expand)
+ depersonalize `I don't think`

> I don't think it would be technically difficult to implement, nor do I think it would add much complexity to the compiler.

This is more apt in the rationale/drawbacks sections.

The reference (AFAIK, for lang RFCs) should be more a language specification and technical description that can be pasted into the actual reference without many changes.